Drop GHC 8.10

amesgen opened the pull request:
The node already uses 9.6, so we can finally drop 8.10 and get rid of a lot of cruft.

I tried to remove everything I could find by grepping for a few ad-hoc regexes; maybe there is more, but we can remove that as we go.

Follow-up work (like upgrading from `Haskell2010` to `GHC2021`) can happen in later PRs as it needs more discussion.

jasagredo commented on .github/workflows/ci.yml:
```diff
@@ -38,7 +38,7 @@ jobs:
       fail-fast: false
       matrix:
-        ghc: ["8.10.7", "9.6.6", "9.10.1", "9.12"]
+        ghc: ["9.6.6", "9.10.1", "9.12"]
```
We could bump 9.6 to 9.6.7 and 9.12 to 9.12.1. But it doesn't need to be now.

amesgen replied:
Will do that in a follow-up PR where I also update haskell.nix :+1: 

(also, there even are 9.10.2 and 9.12.2 already)

jasagredo commented on ouroboros-consensus/ouroboros-consensus.cabal:
```diff
@@ -358,8 +358,6 @@ library
 
   if !impl(ghc >=9.4)
     build-depends: data-array-byte
```
I think this can also be removed?

jasagredo commented:
I also see this:
```
-- ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util/AnchoredFragment.hs
...
    commonPrefix
      | null frags = AF.Empty sharedAnchor
      -- TODO use Foldable1 once all our GHCs support it
      | otherwise = L.foldl1' computeCommonPrefix (toList frags)
```

Not sure if GHC >=9.6.7 supports it

jasagredo commented:
Also worth double-checking this one:
```
-- Ouroboros.Consensus.Protocol.PBFT

#if __GLASGOW_HASKELL__ >= 908
-- GHC is a bit pickier for data family instances, but trying to remove this
-- one forces us to reorganize the Protocol.* modules. TODO eventually.
{-# OPTIONS_GHC -Wno-orphans #-}
#endif
```

jasagredo commented:
And this one:
```
-- Ouroboros.Consensus.HardFork.Combinator.Embed.Unary

{-------------------------------------------------------------------------------
  Defaults (to ease implementation)

  It'd be nicer to use deriving-via here, but we cannot due to a GHC bug
  (resulting in @No family instance for ‘GenTx’@ errors).
  See <https://gitlab.haskell.org/ghc/ghc/issues/13154#note_224287> .
-------------------------------------------------------------------------------}
```

The fix seems to be included in multiple tags dating as far back as 8.10.4: https://gitlab.haskell.org/ghc/ghc/-/commit/cd9b94594440163a1a726300d300f76ff05cd15a

amesgen commented:
Thanks, addressed your comments (apart from https://github.com/IntersectMBO/ouroboros-consensus/pull/1489#issuecomment-2853647494 as we currently still need GHC 9.6), great finds.
